design/one-pager-package-format-v2.md

```diff
@@ -0,0 +1,248 @@
+    # when invoked. Note that this is distinct from the package OCI image.
+    image: example/provider-example:v0.1.0
+  # Optional. Used by Crossplane to ensure any dependencies of a provider are
+  # installed and running before the provider is installed. Unlikely to be used
+  # in practice.
+  dependsOn:
+    # Required. Specifies an OCI image containing a package dependency. This key
+    # may be either 'provider' or 'configuration'. This is sugar; in either case
+    # the package manager determines whether the depencency is really a Provider
+    # or a Configuration by unpacking it and inspecting its kind.
+  - provider: example/provider-uncommon
+    # Required. Will be extended to support version ranges in future, but
+    # currently treated as a specific version tag.
+    version: v0.1.0
+  # Optional. Matching paths will be omitted from the OCI image when building a
+  # package.
+  ignore:
+    # Required. Currently only a path, relative to crossplane.yaml, is allowed.
+  - path: examples/
+```
+
+The above `Provider` example is exhaustive - it contains all supported fields.
+It should be treated as the authoritative example of Provider metadata. Note
+that while it complies with Kubernetes API conventions _it is not a Kubernetes
+custom resource_ - it is never created in an API server and thus no controller
+reconciles it, however it does correspond to a `Provider` custom resource in
+the `packages.crossplane.io` API group. These are two sides of the same coin:
+
+* The `meta.packages.crossplane.io` `Provider` instructs Crossplane how to run
+  the provider, and provides metadata for systems that build on Crossplane -
+  i.e. package registries.
+* The `packages.crossplaneio` `Provider` is submitted to the Crossplane API
+  server in order to declare that a provider should be installed and run. The
+  Crossplane package manager runs a provider by unpacking its package and
+  extracting its `meta.packages.crossplane.io` `Provider` configuration.
+
+Note that the two have different audiences; the `meta.packages.crossplane.io`
+file is authored by the provider maintainer, while the `packages.crossplane.io`
+custom resource is authored by the platform operator.
+
+Contemporary packages contain an `install.yaml` file that allows the package
+author to specify how the package's controller (e.g. the provider) should be
+invoked. This document proposes support for `install.yaml` be removed; packages
+will now only install a specific kind of controller - a Crossplane provider. It
+is possible that both the maintainer of the provider and the platform operator
+will need to influence how the provider is deployed; e.g. by specifying env
+vars, health checks, replica counts, or node selectors. Currently all providers
+use a minimal `install.yaml`; they do not make use of their ability to influence
+how the provider is run. This document recommends that the above `Provider`
```

Comment:
@hasheddan Do you have any thoughts on how we'd handle such a case if we removed the full `install.yaml`? I was thinking we'd add fields to `crossplane.yaml` as appropriate, though we'd want to be fairly confident that we wouldn't eventually grow a full `podSpecTemplate`.

Reply:
The more I think about this the more I feel like most of this configuration should be pushed to "run time" (i.e. "install time") rather than "build time". For instance the IRSA scenario is necessary for some folks when running on EKS but not for the majority of folks. As called out in the package v2 design doc, we are going to need to support providing parameters at install time for a variety of things. I think eventually provider package authors will be able to pick packaging "engines" (i.e. logic in the base image) that allow them to template their deployments based on a bag of input data. I also think that we will want the package manager restrictions to eventually be configurable such that packaged controllers may be able to be installed with varying levels of privilege. All this to say, I think this can be deferred.

Another thing to note here is that some k8s distros like Openshift will do their own mangling of the deployment (particularly the `fsGroup` in this case). This was behavior I considered implementing when we first added support for IRSA, and, ironically, it has actually come back to break our implementation here: https://github.com/crossplane/provider-aws/issues/316

Further reading material [here](https://docs.openshift.com/container-platform/4.1/authentication/managing-security-context-constraints.html).